Light Freshness: ASP.NET - Scaffold + Identity (#37161)
dotnet/AspNetCore.Docs: aspnetcore/security/authentication/add-user-data.md

The project sample is created from a Razor Pages web app, but the instructions are similar for an ASP.NET Core MVC web app.

## Prerequisites

:::moniker range=">= aspnetcore-6.0"

[!INCLUDE [](~/includes/6.0-SDK.md)]

[View or download the sample code](https://github.com/dotnet/AspNetCore.Docs/tree/main/aspnetcore/security/authentication/add-user-data) ([how to download](xref:fundamentals/index#how-to-download-a-sample))

## Create a Razor web app

Create a new project for a Razor web app.

# [Visual Studio](#tab/visual-studio)

1. In Visual Studio, select **File** > **New** > **Project**.

> 1. Enter a name for the project. If you want to the name to match the namespace of the [download sample](https://github.com/dotnet/AspNetCore,Docs/tree/live/aspnetcore/security/authentication/add-user-data) code, enter **WebApp1**.

Reviewer: ```suggestion
1. Enter a name for the project. If you want the name to match the namespace of the [download sample](https://github.com/dotnet/AspNetCore,Docs/tree/live/aspnetcore/security/authentication/add-user-data) code, enter **WebApp1**.
```

Removed an extra "to".

1. Select **ASP.NET Core Web Application** and then select **OK**.

1. In the **Authentication type** dropdown list, select **Individual Accounts**.

> 1. Select **Web Application**  and then select **OK**.

Reviewer: ```suggestion
1. Select **Web Application** and then select **OK**.
```
Removed an extra space.

   Run the following commands in the project directory:

   ```dotnetcli
   dotnet add package Microsoft.VisualStudio.Web.CodeGeneration.Design
   dotnet restore
   ```

1. Run the following command to list the Identity scaffolder options:

   ```dotnetcli
   dotnet aspnet-codegenerator identity -h
   ```

1. In the project folder, run the Identity scaffolder:

   ```dotnetcli
   dotnet aspnet-codegenerator identity -u WebApp1User -fi Account.Register;Account.Manage.Index
   ```

PowerShell uses the semicolon (`;`) as a command separator. When you use PowerShell, escape any semicolons in the file list or put the file list in double quotes (`""`).

---

### Create a migration and check the app

After you prepare the Identity scaffolder, create a migration and check your app. For the following steps, see the detailed instructions in [Migrations, UseAuthentication, and layout](xref:security/authentication/scaffold-identity#migrations-useauthentication-and-layout).

1. Create a migration and update the database.

> 1. Add the `UseAthentication` method to the [Program.cs](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/authentication/add-user-data/samples/6.x/SampleApp/Program.cs#L24) file.

Reviewer: ```suggestion
1. Add the `UseAuthentication` method to the [Program.cs](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/security/authentication/add-user-data/samples/6.x/SampleApp/Program.cs#L24) file.
```
Added missing letter "u" in UseAuthentication".

1. Add the `<partial name="_LoginPartial" />` partial to the layout file. For more information, see [Layout changes](xref:security/authentication/scaffold-identity#layout-changes).

1. Check the app:

   1. Register a user.

   1. Select the new user name (next to the **Logout** link). You might need to expand the window or select the navigation bar icon to show the user name and other links.

   1. Select the **Personal Data** tab.
   
   1. Select **Download** and examine the _PersonalData.json_ file.
   
   1. Select **Delete** and confirm you can delete the signed in user.

## Add custom user data to the Identity database

You need to update the `IdentityUser` derived class with custom properties. If you named the project **WebApp1**, the corresponding file to update is _Areas/Identity/Data/WebApp1User.cs_. Update the file with the following code:

[!code-csharp[](add-user-data/samples/6.x/SampleApp/Areas/Identity/Data/WebApp1User.cs)]

> Properties with the [PersonalDataAttribute](xref:Microsoft.AspNetCore.Identity.PersonalDataAttribute) object are:

Reviewer: ```suggestion
Properties with the [PersonalDataAttribute](xref:Microsoft.AspNetCore.Identity.PersonalDataAttribute) attribute are:
```
This is an attribute, rather than an "object".

---

Follow the instruction in [Migrations, UseAuthentication, and layout](xref:security/authentication/scaffold-identity#migrations-useauthentication-and-layout) to perform the following steps:

* Create a migration and update the database.
* Add `UseAuthentication` to `Startup.Configure`.
* Add `<partial name="_LoginPartial" />` to the layout file.
* Test the app:
  * Register a user

>   * Select the new user name (next to the **Logout** link). You might need to expand the window or select the navigation bar icon to mark the user name and other links visible.

Reviewer: ```suggestion
  * Select the new user name (next to the **Logout** link). You might need to expand the window or select the navigation bar icon to show the user name and other links.
```
"Show" is the correct verb here I think. "Mark ... visible" doesn't convay the intended meaning.
